### DISCUSSION TITLE
Openbook/Getbook <ctext> doesn't clip well

### DISCUSSION THREAD

**Discussion by bobhenl** [Zrips/CMI#4239] [Enhancement]
**Description of issue:**

---
Hi, when I use /openbook rules or /getbook rules, there isn't whole text of rules... It doesn't automatically create more pages, when there's more text on one page (for chat it works normally). So, when I open book, it ends with word:  "smrti, není", but normal custom text continues with this (but these lines aren't in book then):
```
&7povinen ti je vracet
&a- &7Neprosit o op, ranky, předměty
&a- &7Respektovat všechny hráče
&a- &7Respektovat rozhodnutí AT
&a- &7Žádne rasistické zprávy
&a- &7Zákaz módů/hacků
&a- &7Nepsat caps lockem
&a- &7Zákaz 1x1 věží
&a- &7Nestavět v netheru a endu
&a- &7Nevydávat se za člena AT
&a- &7Zákaz sdílení osobních informací jiných hráčů
```
And these lines of this ctext are missing there.... it's happening to every ctext. It should probably normally clip to next page, but it doesn't. And on the beginning of file is <AutoPage>, so where's the problem? Prob. bug

**CONFIG SECTION (DELETE IF NOT RELEVANT):**
https://pastebin.com/AbZSnqs4

---

**Cmi Version (using`/cmi version`):** 8.7.7.4

**Server Type (Spigot/Paperspigot/etc):**  Paper

**Server Version (using `/ver`):** Paper 202, 1.16.3

**Comment by CloudeLecaw** [comment]
Yoo using it wrong
1. I think your lines are toonlong! 
You need to check that a ctext line is NOT longer then a book line if you want to set this as book.
If your lines are long like your are they run out of the booklines ;) 
Yoi should noticed this allready... 

2 work with the format
- <T>at begin of a line
- </T>at the end of a line
<autopage> this is thr first page of your book and ctext
<extpage> for next page of the book and ctext 

Pls check in ctext wiki next time to get more confirm with ctext before creating an error issues


Here is a little example for you with simple 4 Pages ctext book with two lines on each page
I think with this you know now how to reformat your Text and lines. 
Aggain: take allways care and check that each line is NOT longer then a book can have it. Else your line will run out of the bookpage.. 
This also means in most cases if you want a ctext as a book and useable as command to be shown in chat you should create two ctext. One for the book and one for the chatprintout because in chat you can use much longer lines and design the format a bit more comfitable.. Specialy when you got longer lines. 
```
<autopage>
<T>&2this is page</T>
<T>1</T>
<nextpage>
<T>&2this is page</T>
<T>2</T>
<nextpage>
<T>&2this is page</T>
<T>3</T>
<nextpage>
<T>&2this is page</T>
<T>4</T>
```

I hope this could help you :)

**Comment by bobhenl** [comment]
And can't normally book split these longer text lines to more pages? Then there will be so many pages in chat, and this way there's enough pages

**Comment by CloudeLecaw** [comment]
Thats how books work X-D
Thats why i said make one for books and one for chat if u need both
I recomment you to create the bookfile in an Editor and cooy the file to the folder where the ctexts are saved. After reloading cmi new ctexts work fine

**Comment by bobhenl** [comment]
And btw the example you showed me... I know how nextpage works (as you can see my link to pastebin... don't know why do you think that I don't know how it works), I'm using it in my ctexts in chat without problem... just mentioning that it could clip automatically to next page in book, if it's so long. For ex. some plugins also clip lines in GUI items... like Bossshop. So why there isn't limit for MAX characters or lines for the book page, and then it would be clipped? I still think it could be improved, let's see what think Zrips about it.

**Comment by CloudeLecaw** [comment]
yeah maybe it can be improoved to automaticl jump to next line and next page

**Comment by Zrips** [comment]
its not a bug, its just nuisance with how minecraft handles text. Main issue is text width, for example "Neprosit o op, ranky, předměty" and "||||||||||||||||||||||||||||||" has same amount of characters, but one is allot longer. Which means that having one type of text might result in allot less lines then another type of text. So creating pagination for things like this would be pain in but. Especially when this is dependent on users resource pack and what type of font he is using. 
Its simpler for you to adjust lines to fit your text properly by your own needs and if you need to have book and ingame text, you might need to create 2 variations of it. After all, you can show allot more in chat then in a book, so those need to be different in any case.

**Comment by bobhenl** [comment]
OK, I will try to create two variants of text.

**Comment by bobhenl** [comment]
The problem is, that it probably contains some illegal characters, because then EpxloitFixer kicks players due to it... For high packets sent.

**Comment by CloudeLecaw** [comment]
what kind of crazzy code your sending? 
i tried your code and added also some clickable hyperlinks, hovers and command actions to each line and color and all was working fine.

```
&7povinen ti je vracet
&a- &7Neprosit o op, ranky, předměty
&a- &7Respektovat všechny hráče
&a- &7Respektovat rozhodnutí AT
&a- &7Žádne rasistické zprávy
&a- &7Zákaz módů/hacků
&a- &7Nepsat caps lockem
&a- &7Zákaz 1x1 věží
&a- &7Nestavět v netheru a endu
&a- &7Nevydávat se za člena AT
&a- &7Zákaz sdílení osobních informací jiných hráč
```ů

Bookexploifixer is not really needed on higher servers anymore. you said you have Paper 202, 1.16.3 do u have a fix allready.
Well most bookfix plugins tell you its fixed allready in newer versions in their description

**Comment by bobhenl** [comment]
Btw does suggest command works when clicking from book?

**Comment by CloudeLecaw** [comment]
A good question. I never used a wrong command in a book created with ctext. But execute commands on click as player or as console work fine.
Give it a try, i think if it is a cmi command suggestion of right command may work

**Comment by bobhenl** [comment]
For ex. I have this in my book:
<T>&a- &7Další doplňky nalezneš v &5/premium &7(nebo &5/buy&7)</T><H>Klikni pro vykonání příkazu.</H><SC>/premium</SC>

And it doesn't suggest command when clicked. But for ex. normal URL openning works.
Same for <C> etc.

**Comment by CloudeLecaw** [comment]
You need to use ctext!
Look in the wiki how to you ctext :)
https://www.zrips.net/cmi/custom-text/

**Comment by bobhenl** [comment]
It's in book ctext, do you mean it can't be in book?

**Comment by CloudeLecaw** [comment]
It can but it dies not make sense because u see nothing when yoi use buy or it needs to open annother gui (the chat is also a gui) 
Cc and cci comands only make sense on click when its a command that works without needed chat or other gui elements
 Like give an item or tp somewhere or give a permission or give fly.. Or u directly buy an item on click. But only. /buy is not enough 
If the command needs to Show something in chat or needs annother gui it will not work because the bookgui is open.

I would color it, give a nice hover text to the commamd and let it execute on click if its a book gui.

**Comment by bobhenl** [comment]
Oh, and can't it close first bookgui or other guis and then execute ocmmand?

**Comment by CloudeLecaw** [comment]
You got it

**Comment by bobhenl** [comment]
Bow I can't do this for now, or can I?

**Comment by CloudeLecaw** [comment]
No the book will stay open. So only direct execute commands work. The book will stay open after click it
Example what woud work /broadcast Hello (if player has permission or execute it as console in the ctext) 

Also all pastings to Chat do not work on the book gui

**Comment by bobhenl** [comment]
But maybe Zrips can add way to firstly close bookgui and then execute specified commands?

**Comment by CloudeLecaw** [comment]
Maybe :) would be nice if this would be a possible option. 
Specialy for sc and si useage
Ans for commands that need other gui parts after

But for the moment we need to reduce books useage with ctext commands to direct executed commands c, cc or cci

**Comment by bobhenl** [comment]
But still limited, cause when I wanna open another gui via C, I can't :/

**Comment by CloudeLecaw** [comment]
At the moment its not possible. Maybe Zrips can adds this with a feature Update

I like the idea because it would give us much more design option with books with ctexts

**Comment by bobhenl** [comment]
And please look into my sell issue, still having problem there.
